### PR TITLE
OpenSslEngine may lose data if the non-application buffer is small/full

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
If an event occurs which generates non-application data (shutdown, handshake failure, alert generation, etc...) and the non-application buffer in the ByteBuffer BIO is full (or sufficiently small) we may not propagate all data to our peer before tearing down the socket.

Modifications:
- when wrap() detects the outbound is closed, but there is more data pending in the non-application buffers, we must also check if OpenSSL will generate more data from calling SSL_shutdown again
- when wrap() detects a handshakeExcpetion during failure we should check if OpenSSL has any pending data (in addition to the non-application buff) before throwing the handshake exception

Result:
OpenSslEngine more reliably transmits data to the peer before closing the socket.

### CODE REVIEW

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 1]
```diff
@@ -484,6 +492,11 @@ public final SSLEngineResult wrap(
                     bytesProduced = SSL.bioFlushByteBuffer(networkBIO);
 
                     if (bytesProduced > 0 && handshakeException != null) {
+                        // TODO(scott): It is possible that when the handshake failed there was not enough room in the
+                        // non-application buffers to hold the alert. We should get all the data before progressing on.
+                        // However I'm not aware of a way to do this with the OpenSSL APIs.
+                        // See https://github.com/netty/netty/issues/6385.
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: https://github.com/netty/netty/issues/6385

**Comment by Scottmitch** [comment]
@normanmaurer - We already have unit tests :) Its just a reducing the size of the non-application buffer to `1`. However there are other test failures when this is done https://github.com/netty/netty/issues/6385#issuecomment-280078650.

**Comment by normanmaurer** [comment]
@Scottmitch ok got it.. ship it once ci build completes.

**Comment by normanmaurer** [comment]
@Scottmitch rebase on 4.1 ?

**Comment by Scottmitch** [comment]
@normanmaurer - rebased & waiting for CI.

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 2]
```diff
@@ -484,6 +492,11 @@ public final SSLEngineResult wrap(
                     bytesProduced = SSL.bioFlushByteBuffer(networkBIO);
 
                     if (bytesProduced > 0 && handshakeException != null) {
+                        // TODO(scott): It is possible that when the handshake failed there was not enough room in the
```
Comment (netkins) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: INFO Complete the task associated to this TODO comment. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

**Comment by Scottmitch** [comment]
4.1 (4431ad8) 4.0 (2b07949)

**Pull request closed without merging**